// swagger-typed-express-docs/dist/expressRegExUrlParser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseUrlFromExpressRegexp = void 0;
var replacers = {
    prefixUrlSlash: '/^',
    slashBetweenRoutes: '/',
    urlParamString: '(?:([^\\/]+?))',
    endOfRegExpString: '/i',
    endUrlQueryString: '(?=\\/|$)',
    optionalSlash: '\\/?',
    requiredSlash: '\\/',
};
var parseUrlFromExpressRegexp = function (regexpString, params) {
    if (params === void 0) { params = []; }
    var parsedRegExPath = regexpString
        .substr(replacers.prefixUrlSlash.length)
        .slice(0, -replacers.endOfRegExpString.length)
        .slice(0, -replacers.endUrlQueryString.length)
        .split(replacers.urlParamString)
        .flatMap(function (item, index) {
        var _a;
        var isLastItem = params.length > index;
        if (!isLastItem) {
            return [item];
        }
        return [item, ":" + ((_a = params[index]) === null || _a === void 0 ? void 0 : _a.name)];
    })
        .join('')
        .split(replacers.optionalSlash)
        .join('/')
        .split(replacers.requiredSlash)
        .join('/');
    return parsedRegExPath;
};
exports.parseUrlFromExpressRegexp = parseUrlFromExpressRegexp;
