// swagger-typed-express-docs/dist/schemaBuilder.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tNonNullable = exports.tCustomScalar = exports.tList = exports.tObject = exports.tUnion = exports.tAny = exports.tString = exports.tBoolean = exports.tNumber = void 0;
exports.tNumber = {
    type: 'number',
    required: false,
};
exports.tBoolean = {
    type: 'boolean',
    required: false,
};
exports.tString = {
    type: 'string',
    required: false,
};
exports.tAny = {
    type: 'any',
    required: false,
};
var tUnion = function (options) { return ({
    type: 'enum',
    required: false,
    options: options,
}); };
exports.tUnion = tUnion;
var tObject = function (a) { return ({
    type: 'object',
    required: false,
    properties: a,
}); };
exports.tObject = tObject;
var tList = function (items) { return ({
    type: 'array',
    required: false,
    items: items,
}); };
exports.tList = tList;
var tCustomScalar = function (name, conf) { return ({
    name: name,
    type: 'customScalar',
    required: false,
    parser: conf.parser,
    validator: conf.validator,
}); };
exports.tCustomScalar = tCustomScalar;
var tNonNullable = function (a) { return (__assign(__assign({}, a), { required: true })); };
exports.tNonNullable = tNonNullable;
